DBZ-6540 Blogpost about custom signalling/notification channels - part1 (#919)
debezium/debezium.github.io: _posts/2023-06-07-Debezium-signaling-and-notifications.adoc

---
layout: post
title:  Debezium signaling and notifications - Part 1
date:   2023-06-07 11:11:11 +0100
tags: [ debezium, features, notifications, signaling, integration ]
featured: true
author: mfvitale
---
Welcome to this series of articles dedicated to signaling and notifications in Debezium! This post serves as the first installment in the series, where we will introduce the signaling and notification features offered by Debezium and discuss the available channels for interacting with the platform.

In the subsequent parts of this series, we will delve deeper into customizing signaling channels and explore additional topics such as JMX signaling and notifications.

> '''

Reviewer: ```suggestion
+++<!-- more -->+++
```

In today's interconnected world of software applications and systems, the ability to seamlessly integrate with other products is crucial for building robust and efficient solutions.

One of the key challenges in building integrated software ecosystems is establishing effective communication channels between different components and services.

+++<!-- more -->+++

In this context, Debezium now offers comprehensive support for signaling and notification, enabling seamless integration with other product. By leveraging Debezium's signaling and notification mechanisms, users can easily orchestrate actions and stay informed about the status and progress of Debezium.

Debezium signal has been re-designed to be extensible to provide an easy way to add more signaling channels. Furthermore, new out-of-the-box channels has been developed.

We have also added a way to be notified about what is going on inside Debezium through the notification mechanism. Even in this case we provide out-of-the-box channels and an easy way to add more.

> Let's explore the different types of signaling and notification channels available in Debezium, along with their functionalities and benefits.

Reviewer: ```suggestion
In today's interconnected software applications and systems, seamlessly integrating with other products is crucial for building robust and efficient solutions. One of the key challenges in building integrated software ecosystems is establishing effective communication channels between different components and services.

Debezium introduces a comprehensive signaling and notification system providing seamless integration with any third-party solution. Combining the signal and notification systems, you can now orchestrate complex yet efficient pipelines to stay informed about the status and progress of the state managed by Debezium.

The new signaling system has been re-designed to be extensible, providing a simplified approach to adding new, custom signal channels on top of the various implementations available. Let's explore the different signals and notification channels, covering their functionality and describing their benefits.
```
I rewrote this slightly to be more straightforward and reduced the smaller paragraphs into more logical groups.

== Signaling in Debezium
Signaling in Debezium refers to the mechanism through which users can trigger specific actions within the product. It allows users to interact with Debezium and control some behaviors. One notable use case of signaling is the execution of incremental snapshots, which permits to start an ad-hoc snapshot of database.

Debezium provides different signaling channels through which users can send signals to the platform. Let's explore the available signaling channels:

Signaling through Database Tables::

> In earlier versions of Debezium, signaling was only achieved by using a dedicated database table. Users can insert specific records into this table to trigger actions in Debezium connectors. For example, to initiate an incremental snapshot, a record would be inserted into a designated signaling table, prompting the connector to perform the snapshot operation.

Reviewer: ```suggestion
In earlier versions of Debezium, signaling required a dedicated database table where you would insert specific records to trigger Debezium to execute a task. For example, an inserted row could initiate an incremental snapshot, prompting the connector to run a snapshot operation based on the given signal criteria.
```

Signaling through Kafka Topics::
In earlier versions of Debezium, there was available signaling thorough a configurable Kafka topic. The limitation was that it was only available for MySQL with read-only access and GTIDs enabled.
Starting from Debezium version 2.3, Kafka signaling can be used as a signaling mechanism across all connectors. This enhancement provides a more unified approach to signaling, simplifying the integration and allowing for consistent signaling across different databases. Now, users can send signaling messages to a specific Kafka topic, and Debezium will process the signals accordingly.

> Using Kafka for signaling offers several advantages. Firstly, it aligns with the event-driven nature of Debezium, making it a natural fit for managing signals. Additionally, it enables sending signal more easily for user that doesn't have direct access to database.

Reviewer: ```suggestion
In earlier versions of Debezium, you could send signals via a configured Kafka topic; however, this was only available for MySQL using read-only access and global transaction identifiers (GTID) behavior. With Debezium 2.3, the Kafka signal channel is available to all connectors. +
 +
This enhancement provides a simplified integration approach and a unified and consistent approach for signals across all supported Debezium databases. You can send signals to a specific Kafka topic, and Debezium will consume and process that signal as though it originated from the signal table itself. +
 +
Using a Kafka topic for signals provides several advantages. First, it aligns with event-driven design, making it a natural fit with change data capture and Debezium. Additionally, it provides a secure way to send signals to Debezium without necessarily providing the user with direct access to the underlying database.  
```

[NOTE]
===============================

> An important note is that even using Kafka for trigger signal, the incremental snapshot feature still requires the presence of the signaling table, expect for MySQL with GTIDs enabled.

Reviewer: ```suggestion
Even when using the Kafka signal approach, the incremental snapshot feature still requires the presence and use of the signaling table to manage some bookkeeping needed for the incremental snapshot process.
You can only omit the signal table when using MySQL in a read-only way with global transaction identifiers (GTIDs) enabled.
```

===============================

Signaling through File::

> Thanks to our exiting community you can also trigger signals through a simple file. This can be suitable for example when you are not relying on Kafka connect using Debezium server, and you cannot use database signaling.

Reviewer: ```suggestion
You can trigger signals from the file system thanks to a contribution by the Debezium user community. This approach is an excellent alternative when not relying on messaging infrastructures like Kafka combined with Debezium Server or the embedded engine or when you cannot use the database for signals. 
```

Signaling through Java Management Extensions (JMX)::

> With this channel you can send signal calling the signal operation exposed through a dedicated MBean. You can do it by connecting to the MBeanServer with your preferred client. A dedicated post about this feature will follow.

Reviewer: ```suggestion
With this channel, you can send signals calling the signal operation exposed through a dedicated MBean. You can do it by connecting to the MBeanServer with your preferred client. A dedicated post about this feature will follow.
```

Signaling through custom channel::

> As said before, in this release we revisited the signaling mechanism to be extensible. You can provide your own implementation and use it with minimal configuration. A dedicated post about this feature will follow.

Reviewer: ```suggestion
We also redesigned the signaling mechanism to be extensible. You can provide your implementations and use it all via configuration. A dedicated post about this feature will follow.
```

== Notification in Debezium:

> Notification, on the other hand, focuses on obtaining status information and progress updates about Debezium. It provides insights into the execution and completion of tasks such as initial snapshots or incremental snapshots. Debezium generates notifications to keep users informed about the progress of these tasks, facilitating monitoring and decision-making based on real-time information. By leveraging the notification functionality, users can track the state of Debezium and take appropriate actions based on the received notifications.

Reviewer: ```suggestion
Notifications focus on retrieving status and progress information about Debezium and its internal processes. It provides insights into the execution and completion of tasks such as initial snapshots or incremental snapshots. Debezium generates notifications to keep users informed about the progress of these tasks, facilitating monitoring and decision-making based on real-time information. By leveraging the notification functionality, users can track the state of Debezium and take appropriate actions based on the received notifications.
```

> Debezium offers different channels through which notifications can be delivered. Let's explore the out-of-the-box channels:

Reviewer: ```suggestion
Debezium provides a variety of ways to obtain notifications. Let's explore the out-of-the-box channels:
```

Notification through Sink channel::

> This implementation sends notifications through the Connect API to a configured topic. Users can enable this channel and specify the topic name. By publishing notifications to the specified topic, Debezium allows users to consume and process the notifications in a way that suits their needs. Since `Sink` is not technology dependent (i.e. Kafka) it can be also used with Debezium server and its available sinks.

Reviewer: ```suggestion
This implementation sends notifications through the Connect API to a configured topic. Users can enable this channel and specify the topic name. By publishing notifications to the specified topic, Debezium allows users to consume and process notifications in a way that suits their needs. Since `Sink` is agnostic (i.e. not necessarily Kafka), this integrates seamlessly with Debezium Server to deliver notifications to any of its sink adapters.
```

Notification through logs::

> This simple implementations permits to append notifications to the log. Enabling this channel ensures that notifications become part of the application's log stream. This facilitates easy access to notifications for monitoring, debugging, and analysis purposes. Notifications are readily available within the log files, providing a convenient way to track Debezium's progress.

Reviewer: ```suggestion
This channel appends notifications directly to the end of the connector's log. This approach provides convenient access for monitoring, debugging, and analyzing notification details.
```

Notifications through Java Management Extensions (JMX)::

> Enabling this notification channel, Debezium exposes a dedicated MBean with a specific attributes that contains the list of generated notifications. In addition, JMX notifications are exposed as well. Also for this topic, a dedicated post will follow.

Reviewer: ```suggestion
This notification channel exposes a dedicated Debezium MBean with several attributes that contain a list of generated notifications. This channel combines common industry standard monitoring techniques like JMX to react and consume Debezium notifications.  A dedicated post will follow to discuss this in more detail.
```

Notification through custom channel::

> Debezium's notification mechanism is designed to be extensible, enabling users to implement custom channels for delivering notifications in a manner that aligns with their environment. We will cover how to provide a custom notification channel in a dedicated post.

Reviewer: ```suggestion
Debezium's notification mechanism is extensible, allowing users to implement custom channels to deliver notifications using means that best fit their needs. We will cover how to provide a custom notification channel in a dedicated post.
```

== Conclusion

> Signaling and notification are now an important feature of Debezium. Signaling empowers users to interact with Debezium and trigger actions, while notification provides valuable information about Debezium's state and progress.

Reviewer: ```suggestion
Signals and Notifications are now foundational features in Debezium. Signaling empowers users to interact with Debezium and trigger actions, while notification provides valuable information about Debezium's state and progress.
```

> In previous versions of Debezium initiating an incremental snapshot was only possible through the use of a signaling table. Users had to configure a specific table as a signaling mechanism to trigger the incremental snapshot. However, for MySQL with GTIDs enabled, it was possible to utilize Kafka for signaling purposes.

Reviewer: ```suggestion
In previous versions of Debezium, initiating an incremental snapshot was only possible using a signaling table. Users had to configure a specific table as a signaling mechanism to trigger the incremental snapshot. However, for MySQL with GTIDs enabled, it was possible to utilize Kafka for signaling purposes.
```

> With the release of Debezium version 2.3, significant improvements have been introduced. Now, users can utilize Kafka for signaling purposes across all connectors, eliminating the need for a dedicated signaling table. However, it's important to note that for incremental snapshot functionality, the signaling table is still required, unless MySQL with GTIDs enabled is being used.
> 
> These improvements in Debezium's functionality provide users with more streamlined and standardized approaches to signaling and initiating incremental snapshots, enhancing the overall experience of working with Debezium.
> 
> By leveraging the power of signaling and notifications, staying updated with the latest Debezium releases, users can maximize the capabilities of the platform and efficiently capture and process real-time database events.

Reviewer: ```suggestion
With the release of Debezium version 2.3, we have introduced significant improvements to both the signal and notification subsystems.  We've unified several behaviors across connectors and made the entire system extensible, aiming to simplify both custom and future contributed implementations.  We hope this enhances the overall experience of working with Debezium and provides a way to integrate Debezium with other third-party applications and tools seamlessly.  These new and improved features allow you to maximize the capabilities of the Debezium change data capture platform in limitless ways.
```
